pywren/executor.py

```diff
@@ -194,14 +194,14 @@ def map(self, func, iterdata, extra_env=None, extra_meta=None,
                         mod_paths.remove(mod_path)
 
         module_data = create_mod_data(mod_paths)
-        func_str_encoded = wrenutil.bytes_to_b64str(func_str)
+        func_str_encoded = func_str
         #debug_foo = {'func' : func_str_encoded,
         #             'module_data' : module_data}
 
         #pickle.dump(debug_foo, open("/tmp/py35.debug.pickle", 'wb'))
         ### Create func and upload
-        func_module_str = json.dumps({'func' : func_str_encoded,
-                                      'module_data' : module_data})
+        func_module_str = pickle.dumps({'func' : func_str_encoded,
```

Comment:
is there a reason for switching from json to pickle here ? its a bit weird as we already have the func_str pickled if i'm reading the code right

Reply:
we first want to unpickle the modules and get them set up in the `PYTHONPATH` before we unpickle the function, otherwise we will get an error unpickling. Thus the two-stage process.

Reply:
The two stage process is fine -- i was just curious why we changed the second layer or serialization from json to pickle. I might be missing something basic ?

Reply:
We only were using json because we needed wrenhandler, which runs in the lambda-supplied python (2.7 until recently) to write the `foo.py` module files, and  pickle isn't stable cross-python versions. So we needed something independent there, and then internally to that json object we were storing the base64-encoded pickled function -- because pickling results in a binary stream and we then need to base64 it for json. 

In short, pickle is faster to deserialize, more space-efficient, prevents us from having to b64-encode the internal pickled object, and appropriate because the remote `jobrunner` is running the same version of python as the user.